### ISSUE TITLE
False "Unsupported Format" alert on some machines when checking for .der suffix

### ISSUE THREAD

**Issue by jussihi** [rhboot/shim#143]
Hello,

I am building Arch Linux live USB, and as it should be bootable on all BIOS/UEFI/UEFI SecureBoot machines, I installed the [publicly signed shim from AUR](https://aur.archlinux.org/packages/shim-signed/). Everything works fine, I get the MokManager screen when booting, but when I choose my MOK.cer file to be enrolled, MokManager gives me this error:

> Unsupported Format
> Only DER encoded certificate (*.cer/der/crt) is supported

Well, I first thought that I had done some user error here when configuring shim. However, if I stick the exactly same USB stick into my friend's laptop (ASUS TP501U), it works correctly. For the exactly same USB stick it gives me this false error. My laptop is Lenovo X1 Carbon 5th gen (model 20HR006CMX).

This bugs me, is it a bug I've created myself or some bug in shim?

I created the keys with
```
$ openssl req -newkey rsa:2048 -nodes -keyout MOK.key -new -x509 -sha256 -days 3650 -subj "/CN=my Machine Owner Key/" -out MOK.crt
$ openssl x509 -outform DER -in MOK.crt -out MOK.cer
```
as suggested in [ArchWiki](https://wiki.archlinux.org/index.php/Secure_Boot#shim_with_key) and then simply copied the MOK.cer to the USB.

**Comment by ofuangka** [comment]
I'm having the same issue on a Lenovo X1 Carbon 4th gen, using NVMe.  Also, found this post here from someone using a T430s.  Could this be a Lenovo thing?

https://bugzilla.redhat.com/show_bug.cgi?id=1605020

**Comment by jussihi** [comment]
@ofuangka Nice to hear I'm not the only one with this problem. Are you running Arch as well? If so, seems like Arch, Lenovo and NVMe are the common things with us facing the problem/bug. But on the other hand I think that the shim binaries AUR has come straight from Fedora.. So it might not be AUR/Arch related after all...

The bug itself is really weird as it seems to be the `check_der_suffix` function on line 1908 in [MokManager.c](https://github.com/rhboot/shim/blob/master/MokManager.c) that is failing, which is just a simple string comparison check.

**Comment by ofuangka** [comment]
@jussihi Yes, I'm running Arch.  The only reason I brought up NVMe is because I had been able to import MOKs before I upgraded my hard drive (although it's not really a smoking gun because I haven't bothered swapping back yet).

**Comment by jussihi** [comment]
> The only reason I brought up NVMe is because I had been able to import MOKs before I upgraded my hard drive

That definitely means something if you did not update your shim binary in between. Let's wait for some dev to comment on this issue.

**Comment by ofuangka** [comment]
Might not get time to do this until next weekend.  Brought up the change is in case there's some sort of bit order or big/little endian issue.  Also, I couldn't find evidence that the T430s (from the other thread) supports NVMe.  Meanwhile I've been fine just enrolling hashes.  Actually I wonder if the hash comparison uses a different comparison function.

**Comment by ofuangka** [comment]
Actually, I reverted to 13.f4 and was able to import...

**Comment by jussihi** [comment]
> Actually, I reverted to 13.f4 and was able to import...

@ofuangka Ok, then it actually might be the StrnCat<-->StrCpy change in the commit you linked before. Very interesting.

**Comment by ofuangka** [comment]
@jussihi  Just so you know, I found that if you use the older version of shim to import your keys, then upgrade to the newest version, the previously imported keys still work.  Maybe someone can confirm that as a workaround.

**Comment by infomaniac50** [comment]
###TLDR 

Restored my old version of MokManager and key enrollment succeeded.

### Some Context

I'm dual booting an OEM Windows 10 with ArchLinux. I use ArchLinux for work and Windows for SDR and .NET stuff.

My laptop is the Sony VAIO SVE14126CXB and it has the latest BIOS Sony has to offer.

### Story Time

A few days ago I decided to get SecureBoot working.

I signed rEFInd and my kernel with custom local keys, converted the PEM certificate to DER format, and installed everything to the ESP.

rEFInd kept saying SecureBoot was inactive even though the firmware said it was enabled. So I thought maybe it's an old broken Shim just booting whatever without really doing anything.

I updated Shim and MokManager with shim-signed from the AUR. My old Shim was from when I was messing with it in Ubuntu long ago. 

Shim and MokManager are updated now and I reboot. Hmm rEFInd is still saying SecureBoot is inactive. I factory reset SecureBoot because who knows what kind of messed up configs are in there.

I reboot again and this time Shim complains about a security violation. All right, I think to myself, time to enroll my key. I follow the process and now MokManager giving me some nonsense about how refind_local.cer is not DER formatted. What!?

I rebooted and converted the PEM certificate to DER once more; just to make sure I didn't mess up and MokManager still complains. I started Arch one last time and made openssl print out the DER certificate and there was zero complaints. OK. Someone is lying here.

I did some Google fu and found this Github issue. Turns out that I decided to play with SecureBoot 4 days after this issue was reported. (The computer gods have some strange timing.)

I had made backups of the old Ubuntu MokManager before hand. I almost deleted them, but I thought better safe than sorry. I copied the old MokManager over the new one, tried to enroll exact same .cer file that failed before and it worked like a champ. I didn't touch the new Shim and I have not put the new MokManager back yet. I assume MokManager doesn't run unless I start it from rEFInd.

With everything signed and installed I enabled SecureBoot and rebooted. Finally, rEFInd shows up with Windows and ArchLinux as my choices. rEFInd and Windows both say SecureBoot is active. I have not looked in /sys/firmware, but the kernel prints out that thing about EFI stubs and it boots all the way to the userspace LightDM login screen.

I used Arch today for a little bit of research while on battery power. It lasted about an hour but what do you expect from a battery made in 2013.

**Comment by NgoHuy** [comment]
I use mokutil and can import it from my archlinux, but I don't know how to delete hash instead of der file

**Comment by danderson** [comment]
I've just hit this issue as well, using MokManager that ships with Arch's `shim-signed` version 15.f5-1. MokManager thinks my .cer is not valid, whereas OpenSSL and friends are perfectly happy with it.

If it matters, this is on a Dell XPS 15 (9570).

How can I help troubleshoot this further?

**Comment by lcp** [comment]
@NgoHuy Try "mokutil --delete-hash \<hash\>"
@danderson Would you mind to share your cert file?

**Comment by infomaniac50** [comment]
Here's mine for the Sony VAIO SVE14126CXB.

[refind_local.zip](https://github.com/rhboot/shim/files/2457940/refind_local.zip)

**Comment by danderson** [comment]
@lcp Attached. I've included both MOK.crt (PEM encoded) and MOK.cer (DER encoded - I tried to load this file in MokManager). If it would help, I'm also willing to burn this chain of trust and provide the private key, if necessary.

FYI also: I grabbed mmx64.efi from an older release of the Fedora signed shim (https://kojipkgs.fedoraproject.org/packages/shim-signed/13/4/x86_64/shim-x64-13-4.x86_64.rpm), and that MokManager accepted my certificate just fine. I've been able to secure boot via MOKlist since then. So, it looks like something changed between versions 13 and 15, somewhere.

[bad-mok-cert.zip](https://github.com/rhboot/shim/files/2458472/bad-mok-cert.zip)

**Comment by lcp** [comment]
I tested vanilla MokManager from shim-15 with both keys from @infomaniac50 and @danderson in the OVMF virtual machine, and MokManager showed the option to enroll the key as expected. The "view key" option also showed the content of the keys.

mmx64 from fedora (shim-x64-15-8.x86_64.rpm) also works well in the VM.

There is no magic in the suffix check:
https://github.com/rhboot/shim/blob/master/MokManager.c#L1901-L1927
It compares the last 4 characters with the array of suffices and that's it.

Maybe I can write a patch to make MokManager show more debug messages.

**Comment by jussihi** [comment]
@lcp Wow... Hard to believe that a core utility like this can have a bug like this...

**Comment by lcp** [comment]
The patches for gnu-efi:
[gnu-efi-patches.zip](https://github.com/rhboot/shim/files/2468092/gnu-efi-patches.zip)

**Comment by jussihi** [comment]
I guess this can be closed as the patch now seems to be in the gnu-efi upstream too..
